### PR TITLE
UPD: change configuration project

### PR DESCRIPTION
**Pull request by thibaultcordier**
# Description

This configures the pre-commit functionnality: when you prepare a commit, it will automatically run default pre-commit checks and run black and flake8 checks.

Fixes #(issue)

## Type of change

- Configuration files
- pre-commit => change code style

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [x] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [x] Linting passes successfully : `make lint`
- [x] Typing passes successfully : `make type-check`
- [x] Unit tests pass successfully : `make tests`
- [x] Coverage is 100% : `make coverage`
- [x] Documentation builds successfully : `make doc`

### CODE REVIEW

**Comment by thibaultcordier** [comment]
If it looks ok to you, then I merge. What tracks so you can see the real additions :
- .flake8 (https://github.com/scikit-learn-contrib/MAPIE/pull/363/files#diff-6951dbb399883798a226c1fb496fdb4183b1ab48865e75eddecf6ceb6cf46442)
- test.yml (https://github.com/scikit-learn-contrib/MAPIE/pull/363/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88)
- .pre-commit-config.yaml (https://github.com/scikit-learn-contrib/MAPIE/pull/363/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9)
- environment.*.yml
- mypy.ini (https://github.com/scikit-learn-contrib/MAPIE/pull/363/files#diff-6f2d4ba9ca9a357d31014946667b7bed1bfdbc6d2530afc77778fa0a36bee457)
- setup.py (https://github.com/scikit-learn-contrib/MAPIE/pull/363/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7)

To cut a long story short, these modifications automate code formatting during commits, reduce testing during push/pull_request, import only the packages needed according to dev/ci/doc and reduce mapie dependencies to the minimum. The rest of the changes are just formatting the black package.

File: .flake8 [hunk 1]
```diff
@@ -0,0 +1,7 @@
+[flake8]
+exclude = .git, .github, __pycache__ , .vscode, build
+max-line-length = 99
+ignore = E302,E305,W503,E203,E731,E402,E266,E712,F401,F821
```
Comment (vincentblot28) [inline .flake8:4]: E032 : 2 blank k-lines between funcitons --> should be kept
E305 : expected 2 blank lines after function --> should be kept
W503 : Line break occurred before a binary operator --> kept
E203 : white space before ':' --> kept
E402	Module level import not at top of file --> kept
E266	Too many leading '#' for block comment --> kept
E712	Comparison to true should be 'if cond is true:' or 'if cond:' --> kept

F401	Module imported but unused --> kept
F821	Undefined name name --> kept

File: environment.ci.yml [hunk 2]
```diff
@@ -4,9 +4,13 @@ channels:
     - conda-forge
 dependencies:
     - codecov
-    - flake8
+    - flake8==4.0.1
     - mypy
-    - pandas
+    - pytest
     - pytest-cov
-    - scikit-learn
```
Comment (vincentblot28) [inline environment.ci.yml:11]: why sklearn is removed ?

File: environment.dev.yml [hunk 3]
```diff
@@ -4,19 +4,10 @@ channels:
...
     - ipykernel=6.9.0
     - jupyter=1.0.0
-    - mypy=0.941
-    - numpydoc=1.1.0
-    - numpy=1.22.3
-    - pandas=1.3.5
-    - pytest=6.2.5
-    - pytest-cov=3.0.0
-    - python=3.10
-    - scikit-learn
```
Comment (vincentblot28) [inline environment.dev.yml:17]: why sklearn and numpy are removed ?

Comment (vincentblot28) [reply environment.dev.yml:17]: and pytest ?

File: environment.doc.yml [hunk 4]
```diff
@@ -3,12 +3,13 @@ channels:
...
-    - lightgbm=3.2.1
     - numpydoc=1.1.0
-    - pandas=1.3.5
-    - python=3.10
-    - scikit-learn
     - sphinx=4.3.2
     - sphinx-gallery=0.10.1
     - sphinx_rtd_theme=1.0.0
     - typing_extensions=4.0.1
+    - pandas
```
Comment (vincentblot28) [inline environment.doc.yml:11]: why did you remove the versions ?

File: environment.ci.yml (hunk 2 above)
Comment (vincentblot28) [reply environment.ci.yml:11]: same question for the txt env files

File: mapie/classification.py [hunk 5]
```diff
@@ -250,18 +252,18 @@ def _check_depreciated(self) -> None:
...
                 + "the same behavior in the next release.",
-                DeprecationWarning
+                DeprecationWarning,
             )
         if self.method == "cumulated_score":
             warnings.warn(
                 "WARNING: Deprecated method. "
-                + "The method \"cumulated_score\" is outdated. "
-                + "Prefer to use \"aps\" instead to keep "
+                + 'The method "cumulated_score" is outdated. '
```
Comment (vincentblot28) [inline mapie/classification.py:263]: Are we sure we wanted those backslash removed ?

File: mapie/classification.py [hunk 6]
```diff
@@ -424,32 +412,23 @@ def _get_last_index_included(
...
-                ), axis=1
+                    y_pred_proba_cumsum - max_threshold[:, np.newaxis, :], EPSILON
+                ),
+                axis=1,
             )
         else:
             raise ValueError(
-                "Invalid include_last_label argument. "
-                "Should be a boolean or 'randomized'."
+                "Invalid include_last_label argument. " "Should be a boolean or 'randomized'."
```
Comment (vincentblot28) [inline mapie/classification.py:431]: As now the string is on a single line, the quotes have to be updated

File: mapie/classification.py [hunk 7]
```diff
@@ -805,59 +735,36 @@ def _get_last_included_proba(
...
         # get the probability of the last included label
-        y_pred_proba_last = np.take_along_axis(
-            y_pred_proba,
-            y_pred_index_last,
-            axis=1
-        )
+        y_pred_proba_last = np.take_along_axis(y_pred_proba, y_pred_index_last, axis=1)
 
-        zeros_scores_proba_last = (y_pred_proba_last <= EPSILON)
+        zeros_scores_proba_last = y_pred_proba_last <= EPSILON
```
Comment (vincentblot28) [inline mapie/classification.py:759]: this line is easier to understand with the '()'

File: mapie/classification.py [hunk 8]
```diff
@@ -805,59 +735,36 @@ def _get_last_included_proba(
...
         if np.sum(zeros_scores_proba_last) > 0:
             y_pred_proba_last[zeros_scores_proba_last] = np.expand_dims(
-                np.min(
-                    np.ma.masked_less(
-                        y_pred_proba,
-                        EPSILON
-                    ).filled(fill_value=np.inf),
-                    axis=1
-                ), axis=1
+                np.min(np.ma.masked_less(y_pred_proba, EPSILON).filled(fill_value=np.inf), axis=1),
```
Comment (vincentblot28) [inline mapie/classification.py:766]: An example where I think it's less readable after black : here you have 4 numpy functions in 3 lines (expand_dims, min, masked_less, filled), quite hard to understand

File: mapie/classification.py (hunk 5 above)
Comment (thibaultcordier) [reply mapie/classification.py:263]: The quotation marks " and ' have been interchanged. So I think there's no problem.

File: mapie/classification.py (hunk 8 above)
Comment (thibaultcordier) [reply mapie/classification.py:766]: In this case, I should have added pre-commit first, then thought about black.

File: environment.dev.yml (hunk 3 above)
Comment (thibaultcordier) [reply environment.dev.yml:17]: Because of the dependance of mapie installed with `pip install -e`, we also import these packages.

File: environment.ci.yml (hunk 2 above)
Comment (thibaultcordier) [reply environment.ci.yml:11]: Because of the dependance of mapie installed with `pip install -e`, we also import these packages.

File: environment.doc.yml (hunk 4 above)
Comment (thibaultcordier) [reply environment.doc.yml:11]: because there's no incompatibility with the latest version of pandas so far. so there's no reason to constrain the version

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -26,6 +26,7 @@ Please describe the tests that you ran to verify your changes. Provide instructi
 - [ ] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
 - [ ] Linting passes successfully : `make lint`
```
Comment (LacombeLouis) [inline .github/PULL_REQUEST_TEMPLATE.md:27]: What's the difference between `make link` and `make black`? Are we sure that's the correct order?

File: environment.doc.yml (hunk 4 above)
Comment (LacombeLouis) [reply environment.doc.yml:11]: I still feel like we should force a version. We don't want the doc not building for a version problem. (we can perform version updates every now and then?)

File: examples/classification/4-tutorials/plot_main-tutorial-binary-classification.py [hunk 10]
```diff
@@ -79,19 +78,14 @@
 n_classes = 2
 np.random.seed(42)
 X = np.vstack(
-    [
-        np.random.multivariate_normal(center, cov, n_samples)
-        for center, cov in zip(centers, covs)
-    ]
+    [np.random.multivariate_normal(center, cov, n_samples) for center, cov in zip(centers, covs)]
```
Comment (LacombeLouis) [inline examples/classification/4-tutorials/plot_main-tutorial-binary-classification.py:81]: I don't think this is very code reading friendly. Is this `black`?

File: requirements.ci.txt [hunk 11]
```diff
@@ -1,7 +1,10 @@
 codecov
-flake8
+flake8==4.0.1
```
Comment (LacombeLouis) [inline requirements.ci.txt:2]: [question]
What is the usage of the requirements files when we have environment ones?

**Review by LacombeLouis — CHANGES_REQUESTED** [review]
Generally a very clean, good update. Thank you @thibaultcordier.
I think black has some disadvantages, but mainly, it will help clean up the code more than it will make it less readable.

I do wonder whether we should perform the linting before black or not. Please checkout comment.

File: requirements.ci.txt (hunk 11 above)
Comment (thibaultcordier) [reply requirements.ci.txt:2]: This allows the user to choose whether to install with pip or conda.

File: examples/classification/4-tutorials/plot_main-tutorial-binary-classification.py (hunk 10 above)
Comment (thibaultcordier) [reply examples/classification/4-tutorials/plot_main-tutorial-binary-classification.py:81]: Yes, this is `black`

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 9 above)
Comment (thibaultcordier) [reply .github/PULL_REQUEST_TEMPLATE.md:27]: You're right, I'll put it on earlier.

File: examples/classification/4-tutorials/plot_main-tutorial-binary-classification.py (hunk 10 above)
Comment (LacombeLouis) [reply examples/classification/4-tutorials/plot_main-tutorial-binary-classification.py:81]: Ugh, sometimes, black is really not friendly. 😆

**Comment by LacombeLouis** [comment]
Too many changes since this PR. Should be noted when a configuration settings need to be updated.
